cpp/include/cudf/binning/bin.hpp

```diff
@@ -0,0 +1,82 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <cudf/column/column.hpp>
+#include <cudf/column/column_view.hpp>
+#include <cudf/types.hpp>
+
+#include <rmm/mr/device/device_memory_resource.hpp>
+#include <rmm/mr/device/per_device_resource.hpp>
+
+namespace cudf {
+
+/**
+ * @addtogroup binning
+ * @{
+ * @file
+ * @brief APIs for binning values.
+ */
+
+/**
+ * @brief Enum used to define whether or not bins include their boundary points.
+ */
+enum class inclusive { YES, NO };
+
+/**
+ * @brief Labels elements based on membership in the specified bins.
+ *
+ * A bin `i` is defined by `left_edges[i], right_edges[i]`. Whether the edges are inclusive or
+ * not is determined by `left_inclusive` and `right_inclusive`, respectively. If either
+ * `left_edges[i]` or `right_edges[i]` is NULL, no value in `input` can be placed in bin `i`.
```

Comment:
This makes it sound like NULLs are allowed in the edges.

Reply:
Ah that's a holdover from when I was supporting nulls. I'll fix that, but I'll hold off until I get a few more reviews so that I can make all the changes together